router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+	// have a frame that's too big for the MTU, so we have to
+	// fragment it ourself.
+	checkWarn(fragment(dec.Eth, dec.IP, effectivePMTU,
+		func(segFrame []byte) {
+			fwd.aggregate(fwd.aggregatorDFChan, srcName, dstName,
+				segFrame)
+		}))
+}
+
+func (fwd *sleeveForwarder) aggregate(ch chan<- aggregatorFrame, src []byte,
+	dst []byte, frame []byte) {
+	select {
+	case ch <- aggregatorFrame{src, dst, frame}:
+	case <-fwd.finishedChan:
+	}
+}
+
+func fragment(eth layers.Ethernet, ip layers.IPv4, pmtu int,
+	forward func([]byte)) error {
+	// We are not doing any sort of NAT, so we don't need to worry
+	// about checksums of IP payload (eg UDP checksum).
+	headerSize := int(ip.IHL) * 4
+	// &^ is bit clear (AND NOT). So here we're clearing the lowest 3
+	// bits.
+	maxSegmentSize := (pmtu - headerSize) &^ 7
+	opts := gopacket.SerializeOptions{
+		FixLengths:       false,
+		ComputeChecksums: true}
+	payloadSize := int(ip.Length) - headerSize
+	payload := ip.BaseLayer.Payload[:payloadSize]
+	offsetBase := int(ip.FragOffset) << 3
+	origFlags := ip.Flags
+	ip.Flags = ip.Flags | layers.IPv4MoreFragments
+	ip.Length = uint16(headerSize + maxSegmentSize)
+	if eth.EthernetType == layers.EthernetTypeLLC {
+		// using LLC, so must set eth length correctly. eth length
+		// is just the length of the payload
+		eth.Length = ip.Length
+	} else {
+		eth.Length = 0
+	}
+	for offset := 0; offset < payloadSize; offset += maxSegmentSize {
+		var segmentPayload []byte
+		if len(payload) <= maxSegmentSize {
+			// last one
+			segmentPayload = payload
+			ip.Length = uint16(len(payload) + headerSize)
+			ip.Flags = origFlags
+			if eth.EthernetType == layers.EthernetTypeLLC {
+				eth.Length = ip.Length
+			} else {
+				eth.Length = 0
+			}
+		} else {
+			segmentPayload = payload[:maxSegmentSize]
+			payload = payload[maxSegmentSize:]
+		}
+		ip.FragOffset = uint16((offset + offsetBase) >> 3)
+		buf := gopacket.NewSerializeBuffer()
+		segPayload := gopacket.Payload(segmentPayload)
+		err := gopacket.SerializeLayers(buf, opts, &eth, &ip,
+			&segPayload)
+		if err != nil {
+			return err
+		}
+
+		forward(buf.Bytes())
+	}
+	return nil
+}
+
+func frameTooBig(frame []byte, effectivePMTU int) bool {
+	// We capture/forward complete ethernet frames. Therefore the
+	// frame length includes the ethernet header. However, MTUs
+	// operate at the IP layer and thus do not include the ethernet
+	// header. To put it another way, when a sender that was told an
+	// MTU of M sends an IP packet of exactly that length, we will
+	// capture/forward M + EthernetOverhead bytes of data.
+	return len(frame) > effectivePMTU+EthernetOverhead
+}
+
+func (fwd *sleeveForwarder) ControlMessage(msg []byte) {
+	select {
+	case fwd.specialChan <- specialFrame{nil, msg}:
+	case <-fwd.finishedChan:
+	}
+}
+
+func (fwd *sleeveForwarder) Close() {
+	sleeve := fwd.sleeve
+	sleeve.lock.Lock()
+	if sleeve.forwarders[fwd.remotePeer.Name] == fwd {
+		delete(sleeve.forwarders, fwd.remotePeer.Name)
```

Comment:
We have separate `addForwarder` and `lookupForwarder` methods on `Sleeve` - why not `removeForwarder` too?